eth_account/messages.py

```diff
@@ -56,6 +62,102 @@ class SignableMessage(NamedTuple):
     body: bytes  # aka "data to sign"
 
 
+@dataclass(iter=True, slots=True)
+class EIP712Type:
+    @property
+    def type(self) -> str:
+        return self.__class__.__name__
+
+    def field_type(self, field: str) -> str:
+        typ = self.__annotations__[field]
+
+        if isinstance(typ, str):
+            if not is_encodable_type(typ):
+                raise ValidationError(
+                    f"'{field}: {typ}' is not a valid ABI type"
+                )
+
+            return typ
+
+        elif issubclass(typ, EIP712Type):
+            return typ.type
+
+        else:
+            raise ValidationError(
+                f"'{field}' type annotation must either be a subclass of "
+                f"`EIP712Type` or valid ABI Type string, not {typ.__class__.__name__}"
+            )
+
+    def types(self) -> dict:
+        types = {}
+        types[self.type] = []
+
+        for field in fields(self):
+            value = getattr(self, field)
+            if isinstance(value, EIP712Type):
+                types[self.type].append({"name": field, "type": value.type})
+                types.update(value.types())
+            else:
+                types[self.type].append({"name": field, "type": self.field_type(field)})
+
+        return types
+
+    @property
+    def data(self) -> dict:
+        return as_dict(self)  # NOTE: Handles recursion
+
+
+# TODO: Make type of EIP712Message a subtype of SignableMessage somehow
```

Comment:
If this proves *difficult* maybe just a convenience API like `MyEIP712Message.as_signable_message()`

Reply:
In fact, whether or not it is difficult to subclass, I'd consider the "emit signable message" approach to be preferable.

Reply:
It is more explicit

Reply:
I think the intent is to keep the signing API dead simple, and to have `EIP712Message.as_signable_message()` emit that simpler `SignableMessage` API.

Reply:
Actually, all of this about creating a signed message seems like the wrong direction. The pattern in eth-account is to use the signable-message-building methods. In fact there is already one for 712:
https://github.com/ethereum/eth-account/blob/c4bc769039fdacb8eddf26c84640af84651a9da2/eth_account/messages.py#L114

I'm signing off soon, but I'll come back and take a look to see if we can find another approach. Duplicate ways of building 712 signatures doesn't seem like the right approach, right now.

Reply:
Could turn this into another library, just not really a huge fan of how the current API works because it really doesn't fit well with how I'd want to use it, which is more like creating a struct and leveraging that several times. This approach using dataclasses seemed much cleaner from an end user perspective, but I could definitely see how it might better live in another library or `web3py`

Reply:
Yeah, I think it's very reasonable for it to live in its own library. As long as the library can generate a `SignableMessage`, then it should be trivial to just use the existing eth-account APIs. It doesn't really need to live in here.

Reply:
Will look into this, when I have more time.

Reply:
Cool. Once a separate 712 implementation looks good, I think we would probably just strip out the one already in here.

Reply:
Yeah this one is pretty awkward to use right now